### PR TITLE
[Messenger] Add new Google Pub/Sub transport

### PR DESCRIPTION
**Pull request by richardhj**
| Q             | A
| ------------- | ---
| Branch?       | 5.x
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | N/A
| License       | MIT
| Doc PR        | TODO

This PR adds a Google Pub/Sub transport bridge for the Messenger component.

Google Pub/Sub is different from other transports like Amazon SQS. Instead of queues, Pub/Sub has topics and subscriptions. One publishes a message to a topic and it’s delivered once per subscription. This comes with both higher flexibility and higher complexity.  
In practice, you can either set-up one transport in Symfony with one topic and one subscription or multiple transports with different topics and subscriptions. You can also define transports that only receive messages from one topic but cannot publish. See the snippet below for some examples. I found [this documentation about 1:n pub/sub systems](https://cloud.google.com/pubsub/docs/building-pubsub-messaging-system) helpful.

##### Authentication

To access the Pub/Sub service, you must first set up authentication by creating a service account. Follow [this documentation](https://cloud.google.com/pubsub/docs/reference/libraries#cloud-console). For the messenger, assign at least the roles "Pub/Sub viewer" and ("Pub/Sub publisher" and/or "Pub/Sub subscriber"). Place the JSON file in your Symfony project.

##### Symfony config
```yaml
# /config/packages/messenger.yml

framework:
  messenger:
    transports:
      async:
        dsn: 'gps://default/my-awesome-project?topic[name]=my-topic-1&subscription[name]=my-subscription-1'
        options:
          # Define the access key as content (must be an array from the json) OR absolute path. Can also be configured via DSN
          keyFile: '%env(json:file:GCP_KEY_FILE)%'
          keyFilePath: '%kernel.project_dir%/config/gcp/key-c434604cd07b.json'

      can_only_send: 'gps://default/my-awesome-project?topic[name]=my-topic-1'
      can_only_receive: 'gps://default/my-awesome-project?subscription[name]=my-subscription-1'
```

#### TODO:

* [ ] Tests (incl. integration test with emulator)

_____

Looking forward to some reviews. Eventually from people having experience with Google Pub/Sub.
Looking good? Then I'll happy to add the tests.

### CODE REVIEW

File: src/Symfony/Component/Messenger/Bridge/GooglePubSub/Transport/GooglePubSubReceiver.php [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+            return;
+        }
+
+        try {
+            $envelope = $this->serializer->decode([
+                'body' => $gpsEnvelope['body'],
+                'headers' => $gpsEnvelope['headers'],
+            ]);
+        } catch (MessageDecodingFailedException $exception) {
+            $this->subscription->delete($gpsEnvelope['id']);
```
Comment (richardhj) [inline src/Symfony/Component/Messenger/Bridge/GooglePubSub/Transport/GooglePubSubReceiver.php:58]: Should reject the message instead.

File: src/Symfony/Component/Messenger/Bridge/GooglePubSub/composer.json [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    },
+    "autoload": {
+        "psr-4": { "Symfony\\Component\\Messenger\\Bridge\\GooglePubSub\\": "" },
+        "exclude-from-classmap": [
+            "/Tests/"
+        ]
+    },
+    "minimum-stability": "dev",
+    "extra": {
+        "branch-version": "5.2"
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Messenger/Bridge/GooglePubSub/composer.json:38]: 5.3 :)

File: src/Symfony/Component/Messenger/Bridge/GooglePubSub/Transport/GooglePubSubTransportFactory.php [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+    public function createTransport(string $dsn, array $options, SerializerInterface $serializer): TransportInterface
+    {
+        unset($options['transport_name']);
+
+        return new GooglePubSubTransport(Connection::fromDsn($dsn, $options), $serializer);
+    }
+
+    public function supports(string $dsn, array $options): bool
+    {
+        return 0 === strpos($dsn, 'gps://');
```
Comment (Guikingone) [inline src/Symfony/Component/Messenger/Bridge/GooglePubSub/Transport/GooglePubSubTransportFactory.php:32]: What about `return 0 === strpos($dsn, 'gps://') || 0 === strpos($dsn, 'gcp://')`?

Comment (richardhj) [reply src/Symfony/Component/Messenger/Bridge/GooglePubSub/Transport/GooglePubSubTransportFactory.php:32]: There is also Google Pub/Sub Lite which uses a different API and thus is not compatible. GCP (= Google Cloud Platform) would've been too generic for me.

File: src/Symfony/Component/Messenger/Bridge/GooglePubSub/Transport/GooglePubSubSender.php [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Messenger\Bridge\GooglePubSub\Transport;
+
+use AsyncAws\Core\Exception\Http\HttpException;
```
Comment (derrabus) [inline src/Symfony/Component/Messenger/Bridge/GooglePubSub/Transport/GooglePubSubSender.php:14]: This doesn't feel right.

File: src/Symfony/Component/Messenger/Bridge/GooglePubSub/composer.json [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+            "name": "Symfony Community",
+            "homepage": "https://symfony.com/contributors"
+        }
+    ],
+    "require": {
+        "php": ">=7.2.5",
+        "ext-json": "*",
+        "google/cloud-pubsub": "^1.0",
+        "symfony/messenger": "^4.3|^5.0",
+        "symfony/service-contracts": "^1.1|^2"
```
Comment (derrabus) [inline src/Symfony/Component/Messenger/Bridge/GooglePubSub/composer.json:23]: You're not using the service contracts anywhere, are you? Please only add dependencies you actually use (same for the `require-dev` part).

File: src/Symfony/Component/Messenger/Bridge/GooglePubSub/Transport/Connection.php [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+        }
+
+        $query = [];
+        if (isset($parsedUrl['query'])) {
+            parse_str($parsedUrl['query'], $query);
+        }
+
+        // Check for unsupported keys in options
+        $optionsExtraKeys = array_diff(array_keys($options), array_keys(self::DEFAULT_OPTIONS));
+        if (0 < \count($optionsExtraKeys)) {
```
Comment (derrabus) [inline src/Symfony/Component/Messenger/Bridge/GooglePubSub/Transport/Connection.php:92]:
```suggestion
        if (!empty($optionsExtraKeys)) {
```

File: src/Symfony/Component/Messenger/Bridge/GooglePubSub/Transport/Connection.php [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+
+        $options       = $query + $options + self::DEFAULT_OPTIONS;
+        $configuration = [
+            'topic' => $options['topic'],
+            'subscription' => $options['subscription'],
+            'auto_setup' => (bool) $options['auto_setup'],
+        ];
+
+        $clientConfiguration = [
+            'projectId' => ltrim($parsedUrl['path'] ?? '', '/') ?? self::DEFAULT_OPTIONS['projectId'],
```
Comment (derrabus) [inline src/Symfony/Component/Messenger/Bridge/GooglePubSub/Transport/Connection.php:110]: This doesn't look right. The return value of `ltrim` is never `null`, so the second null coalescing operator won't do anything.

**Review by derrabus — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/Messenger/Bridge/GooglePubSub/composer.json [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+        },
+        {
+            "name": "Symfony Community",
+            "homepage": "https://symfony.com/contributors"
+        }
+    ],
+    "require": {
+        "php": ">=7.2.5",
+        "ext-json": "*",
+        "google/cloud-pubsub": "^1.0",
```
Comment (Nyholm) [inline src/Symfony/Component/Messenger/Bridge/GooglePubSub/composer.json:21]: This will download the following packages: 

```
google/protobuf
google/common-protos
ralouphie/getallheaders
psr/http-message
guzzlehttp/psr7
guzzlehttp/promises
grpc/grpc 
firebase/php-jwt
psr/http-client
guzzlehttp/guzzle
google/auth 
google/grpc-gcp
google/gax 
monolog/monolog 
rize/uri-template 
google/cloud-core
google/cloud-pubsub
```

All that for sending 4 HTTP request. Could we not depend on an external package like this? Maybe do the HTTP requests with the Symfony HTTP client?

Comment (richardhj) [reply src/Symfony/Component/Messenger/Bridge/GooglePubSub/composer.json:21]: You are right. The official SDK surely is the most convenient approach but it comes with Guzzle etc. Using the HttpClient will work as well.

Comment (Nyholm) [reply src/Symfony/Component/Messenger/Bridge/GooglePubSub/composer.json:21]: Awesome. Thank you

File: src/Symfony/Component/Messenger/Bridge/GooglePubSub/Transport/Connection.php (hunk 6 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/Messenger/Bridge/GooglePubSub/Transport/Connection.php:92]: (we try to avoid using empty() when possible)

File: src/Symfony/Component/Messenger/Bridge/GooglePubSub/composer.json (hunk 8 above)
Comment (fabpot) [reply src/Symfony/Component/Messenger/Bridge/GooglePubSub/composer.json:21]: This is a critical point and the rule we have for third party providers: we never depend on the official SDK if we can rely on the HttpClient component (and if we must depend on the official SDK, then it's not for Symfony core).

Comment (richardhj) [reply src/Symfony/Component/Messenger/Bridge/GooglePubSub/composer.json:21]: Thanks for the clarification. Will update soon 👍

File: src/Symfony/Component/Messenger/Bridge/GooglePubSub/Transport/Connection.php (hunk 6 above)
Comment (derrabus) [reply src/Symfony/Component/Messenger/Bridge/GooglePubSub/Transport/Connection.php:92]: I see, all right then.

File: src/Symfony/Component/Messenger/Bridge/GooglePubSub/Transport/GooglePubSubReceiver.php [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+
+            $attributes = $message->attributes();
+            if (isset($attributes[self::MESSAGE_ATTRIBUTE_NAME])) {
+                $headers = json_decode($attributes[self::MESSAGE_ATTRIBUTE_NAME], true);
+                unset($attributes[self::MESSAGE_ATTRIBUTE_NAME]);
+            }
+
+            $headers += $attributes;
+
+            $this->buffer[] = [
```
Comment (jderusse) [inline src/Symfony/Component/Messenger/Bridge/GooglePubSub/Transport/GooglePubSubReceiver.php:138]: if you prefetch messages,. you should setup a mechanism to requeue messages on _destruct. 
Worker can decide to stop fetching message at any time.

see Sqs connection for example.

File: src/Symfony/Component/Messenger/Bridge/GooglePubSub/Transport/GooglePubSubReceiver.php [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+    private function getNewMessages(): \Generator
+    {
+        $this->fetchMessages();
+
+        yield from $this->getPendingMessages();
+    }
+
+    private function fetchMessages(): void
+    {
+        foreach ($this->subscription->pull() as $message) {
```
Comment (jderusse) [inline src/Symfony/Component/Messenger/Bridge/GooglePubSub/Transport/GooglePubSubReceiver.php:127]: By default, the SDK returns 1000 messages.
It sound like a lot to store in memory and to ack before the deadline.
I would make it parameterizable.

File: src/Symfony/Component/Messenger/Bridge/GooglePubSub/Transport/GooglePubSubReceiver.php (hunk 1 above)
Comment (jderusse) [inline src/Symfony/Component/Messenger/Bridge/GooglePubSub/Transport/GooglePubSubReceiver.php:58]: this deletes the subscription. did you mean "acknowledge"?

File: src/Symfony/Component/Messenger/Bridge/GooglePubSub/composer.json (hunk 8 above)
Comment (Guikingone) [reply src/Symfony/Component/Messenger/Bridge/GooglePubSub/composer.json:21]: When it comes to using HttpClient, don't forget that the official SDK use shortcuts and hooks during authentication process (mostly when it calls the API endpoints), IMO it could be a good idea to use a scoped client 🙂

File: src/Symfony/Component/Messenger/Bridge/GooglePubSub/Transport/Connection.php [hunk 11]
```diff
@@ -0,0 +1,188 @@
...
+        ],
+
+        // PubSub subscription
+        'subscription' => [
+            'name' => null
+        ],
+
+        // Internal
+        'auto_setup' => true,
+    ];
```
Comment (Guikingone) [inline src/Symfony/Component/Messenger/Bridge/GooglePubSub/Transport/Connection.php:51]: I don't know for others but maybe it could be a good idea to use `OptionsResolver` to define the options of this transport, this way, we can easily inspect and validate the submitted options when building it?

**Comment by richardhj** [comment]
I made efforts to check what it means to get rid of the [google/cloud-pubsub](https://github.com/googleapis/google-cloud-php-pubsub/) dependency.

All the GoogleAPI packages are super nested and require many superfluous dependencies (e.g., required firebase components that do not belong to google core; e.g., required Guzzle and Monolog packages instead of using the PSR 17 and PSR 3 abstractions etc.pp.). It kind of feels the SDK needs an overhaul, but it also lacks alternatives.

First of all, making the actual requests to the Google-API is simple and can be done solely with the Http-Client.

The tricky part is the authentication. You usually create a service account in the cloud console and get a key file in exchange. The SDK uses the key file to make an API request to gather an OAuth token. This token then can be used as a Bearer token for the subsequent API requests.
We would need to rebuild the whole service: https://github.com/googleapis/google-auth-library-php/blob/master/src/OAuth2.php#L526

An alternative would be to use API keys, but the API keys do not allow fine-granular access control on the API-level.

@Guikingone (by coincidence?) opened an issue to ask for an HTTP-client abstraction: googleapis/google-auth-library-php#316. Let's see if this will get the stone rolling.

**TLDR:**
- Using the current Google SDK is no option because it comes with too many opinionated dependencies.
- Implementing the auth part ourselves is too extensive to be part of this Symfony bridge.

**Comment by nicolas-grekas** [comment]
Is the oauth part generic enough to deserve being added to the http client component?

**Comment by Guikingone** [comment]
What about using https://github.com/thephpleague/oauth2-client to authenticate the bridge?

**Review by jderusse — COMMENTED** [review]
I all other transports, the communication with the backend is performed by the `Connection` class.

In this PR, this is performed directly by the Subscriber, the Receiver.

I think for consistency, this transport should follow the same pattern.

I think, for consistency, this

File: src/Symfony/Component/Messenger/Bridge/GooglePubSub/Transport/GooglePubSubReceiver.php [hunk 12]
```diff
@@ -0,0 +1,146 @@
...
+
+        $this->subscription->acknowledge(new Message([], ['ackId' => $stamp->getAckId()]));
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function reject(Envelope $envelope): void
+    {
+        $this->ack($envelope);
```
Comment (marforon) [inline src/Symfony/Component/Messenger/Bridge/GooglePubSub/Transport/GooglePubSubReceiver.php:81]: This has to be wrong. You can NACK message by modifying ack deadline to 0 as it's documented here https://cloud.google.com/pubsub/docs/reference/rest/v1/projects.subscriptions/modifyAckDeadline. If you do ack on reject, failed message won't be redelivered even if you have set up retry policy on subscription.

Comment (richardhj) [reply src/Symfony/Component/Messenger/Bridge/GooglePubSub/Transport/GooglePubSubReceiver.php:81]: Ok, that's a great point! Thanks for bringing this up.
So basically `modifyAckDeadline(0)` will allow other subscribers to handle the message while `ack()` will permanently remove the message (which is not what we want in a Pub/Sub setting)

**Comment by fabpot** [comment]
Closing this PR as there is no more activity and as we won't add a notifier that depends on an external SDK. We only want notifiers that can only use HttpClient.
